### PR TITLE
Np 1204 refactor to accomodate user update

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: src/main/java/no/unit/nva/database/DatabaseServiceImpl.java [hunk 1]
```diff
@@ -57,61 +65,126 @@ public DatabaseServiceImpl(DynamoDBMapper mapper) {
...
-            .forEach(role -> mapper.save(roleDto.toRoleDb()))
-            .orElseThrow(failure -> new InvalidInputRoleException(INVALID_ROLE_ERROR));
+
+        validate(roleDto);
+        checkRoleDoesNotExist(roleDto);
+        mapper.save(roleDto.toRoleDb());
+    }
+
+    @Override
+    public void updateUser(UserDto queryObject)
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/database/DatabaseServiceImpl.java:93]: The only functionality tested  is updating existing user. More tests will be added for this functionality later.

File: src/test/java/no/unit/nva/handlers/AddUserTest.java [hunk 2]
```diff
@@ -133,22 +129,6 @@ public void processInputThrowsConflictExceptionWhenInputUserDoesNotHaveUsername(
         assertThat(response.getStatusCode(), is(equalTo(HttpStatus.SC_BAD_REQUEST)));
     }
 
-    @DisplayName("processInput() throws RuntimeException when database service throws unexpected exception "
-        + "when getting user ")
-    @Test
-    public void processInputThrowsUnexpectedExceptionWhenDatabaseServiceThrowsUnexpectedExceptionWhenReadingUser()
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/handlers/AddUserTest.java:139]: This test was testing the case where the valid data sent to the database are read again and now they are invalid. This basically tests the rare case where there is an error "on the wire" when getting the data.   There is little point on testing this extreme case. The only reason this test was here was for test coverage.

File: src/main/java/no/unit/nva/database/DatabaseServiceImpl.java [hunk 3]
```diff
@@ -57,65 +64,130 @@ public DatabaseServiceImpl(DynamoDBMapper mapper) {
...
-
-        UserDto user = userSearchResult
-            .stream()
-            .map(attempt(UserDto::fromUserDb))
-            .map(attempt -> attempt.orElseThrow(this::unexpectedException))
-            .collect(SingletonCollector.collectOrElse(null));
-
-        return Optional.ofNullable(user);
+    public UserDto getUser(UserDto queryObject) throws InvalidEntryInternalException, NotFoundException {
+        Optional<UserDto> resultOpt = getUserAsOptional(queryObject);
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/database/DatabaseServiceImpl.java]: return getUserAsOptional(queryObject)
            .orElseThrow(() -> new NotFoundException(USER_NOT_FOUND_MESSAGE + queryObject.getUsername()));

File: src/main/java/no/unit/nva/database/DatabaseServiceImpl.java [hunk 4]
```diff
@@ -57,65 +64,130 @@ public DatabaseServiceImpl(DynamoDBMapper mapper) {
...
-                .stream()
-                .map(attempt(RoleDto::fromRoleDb))
-                .map(attempt -> attempt.orElseThrow(this::unexpectedException))
-                .collect(SingletonCollector.collectOrElse(null));
-        return Optional.ofNullable(retrievedRole);
+        return convertQueryResultToOptionalRole(queryObject, searchRoleByNameResult);
+    }
+
+    @Override
+    public Optional<UserDto> getUserAsOptional(UserDto queryObject) throws InvalidEntryInternalException {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/database/DatabaseServiceImpl.java:123]: I don't think you actually need to pass an Optional…ever

File: src/main/java/no/unit/nva/database/DatabaseServiceImpl.java [hunk 5]
```diff
@@ -57,65 +64,130 @@ public DatabaseServiceImpl(DynamoDBMapper mapper) {
...
 
     @Override
-    public Optional<RoleDto> getRole(RoleDto queryObject) throws InvalidRoleInternalException {
+    public RoleDto getRole(RoleDto queryObject) throws NotFoundException, InvalidEntryInternalException {
+        return getRoleAsOptional(queryObject)
+            .orElseThrow(() -> new NotFoundException(ROLE_NOT_FOUND_MESSAGE + queryObject.getRoleName()));
+    }
+
+    @Override
+    public Optional<RoleDto> getRoleAsOptional(RoleDto queryObject) throws InvalidEntryInternalException {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/database/DatabaseServiceImpl.java:113]: I don't think you actually need to pass an Optional…ever

File: src/main/java/no/unit/nva/handlers/AddUserHandler.java [hunk 6]
```diff
@@ -17,8 +18,9 @@
 
 public class AddUserHandler extends HandlerWithEventualConsistency<UserDto, UserDto> {
 
-
     public static final String SYNC_ERROR_MESSAGE = "Error while trying to retrieve saved user:";
+    public static final String PUBLIC_MESSAGE_FOR_INTERNAL_CONSISTENCY_PROBLEMS = "Problem with the data has occured";
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/handlers/AddUserHandler.java]: A problem

File: src/main/java/no/unit/nva/handlers/AddUserHandler.java [hunk 7]
```diff
@@ -17,8 +18,9 @@
...
+    public static final String INCOSISTEND_DATA_ERROR = "Inconsistent data in the database.";
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/handlers/AddUserHandler.java]: Inconsist-ent

File: src/main/java/no/unit/nva/model/UserDto.java [hunk 8]
```diff
@@ -13,22 +13,22 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class UserDto implements WithCopy<UserDto.Builder>, JsonSerializable {
+public class UserDto implements WithCopy<UserDto.Builder>, JsonSerializable, Validable {
 
     public static final String MISSING_FIELD_ERROR = "Invalid User. Missing obligatory field: ";
     public static final String ERROR_DUE_TO_INVALID_ROLE =
         "Failure while trying to create user with role without role-name";
+    public static final String INVALID_USER_ERROR_MESSAGE = "Invalid user. User should have non empty username.";
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/model/UserDto.java]: non-empty

File: src/test/java/no/unit/nva/database/DatabaseServiceTest.java [hunk 9]
```diff
@@ -91,25 +90,34 @@ public void addUserShouldSaveUserWithoutRoles()
...
+    public void updateUserUpdatesExistingUserWithInputUserWhenInputUserIsValid()
+        throws ConflictException, InvalidEntryInternalException, NotFoundException,
+               InvalidInputException {
+        UserDto existingUser = createSampleUserAndAddUserToDb(SOME_USERNAME, SOME_INSTITUTION, SOME_ROLE);
+        UserDto expectedUser = alterUser(existingUser);
+
+        db.updateUser(expectedUser);
+    }
+
+    private UserDto alterUser(UserDto existingUser) throws InvalidEntryInternalException {
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/database/DatabaseServiceTest.java]: cloneAndChangeRole?

**Review by brinxmat — CHANGES_REQUESTED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**